pkg/image/apis/image/validation/whitelist/whitelister.go

```diff
@@ -0,0 +1,268 @@
+package whitelist
+
+import (
+	"fmt"
+	"net"
+	"reflect"
+	"strings"
+
+	"github.com/golang/glog"
+
+	kerrutil "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	serverapi "github.com/openshift/origin/pkg/cmd/server/api"
+	imageapi "github.com/openshift/origin/pkg/image/apis/image"
+	stringsutil "github.com/openshift/origin/pkg/util/strings"
+)
+
+// WhitelistTransport says whether the associated registry host shall be treated as secure or insecure.
+type WhitelistTransport string
+
+const (
+	WhitelistTransportAny      WhitelistTransport = "any"
+	WhitelistTransportSecure   WhitelistTransport = "secure"
+	WhitelistTransportInsecure WhitelistTransport = "insecure"
+)
+
+// RegistryWhitelister decides whether given image pull specs are allowed by system's image policy.
+type RegistryWhitelister interface {
+	// AdmitHostname returns error if the given host is not allowed by the whitelist.
+	AdmitHostname(host string, transport WhitelistTransport) error
+	// AdmitPullSpec returns error if the given pull spec is allowed neither by the whitelist nor by the
+	// collected whitelisted pull specs.
+	AdmitPullSpec(pullSpec string, transport WhitelistTransport) error
+	// AdmitDockerImageReference returns error if the given reference is allowed neither by the whitelist nor
+	// by the collected whitelisted pull specs.
+	AdmitDockerImageReference(ref *imageapi.DockerImageReference, transport WhitelistTransport) error
+	// WhitelistRegistry extends internal whitelist for additional registry domain name. Accepted values are:
+	//  <host>, <host>:<port>
+	// where each component can contain wildcards like '*' or '??' to match wide range of registries. If the
```

Comment:
Why are we supporting `??`?

Reply:
I'm just reusing what is right now being used in [ValidateRegistryAllowedForImport](https://github.com/openshift/origin/pull/17783/files/ead80b88ca74703a83481d59ed24897e54e2f609#diff-f7c62fc51d70d941e39d08ad02da433bL302) ([IsWildcardMatch](https://github.com/miminar/origin/blob/ead80b88ca74703a83481d59ed24897e54e2f609/pkg/util/strings/wildcard.go#L6)). Somebody may have already started using it with registries allows for import.